File: GoInterviewQuestion/360/search/69. x 的平方根.go
package search


func mySqrt(x int) int {
	if x == 1 {
		return 1
	}
	//二分查找的思想
	i, j := 1, x/2
	for i <= j {
		mid := (i+j)/2
		if mid*mid > x {
			j = mid - 1
		} else if mid * mid < x {
			i = mid + 1
		} else {
			return mid
		}
	}
	return i - 1
}

//i < j的解法
func mySqrt1(x int) int {
	if x <= 1 {
		return x
	}
	//二分查找的思想
	i, j := 1, x/2
	for i < j {
		//i = mid, 就需要中间值偏向j
		mid := (i+j+1)/2
		if mid*mid > x {
			j = mid - 1
		} else if mid * mid < x {
			i = mid
		} else {
			return mid
		}
	}
	return i
}

File: PythonLeetcode/BinarySearch/69. x 的平方根.py
import sys


class Solution:
    @staticmethod
    def my_sqrt(x: int) -> int:
        i, j = 1, x
        while i < j:
            mid = int((i + j + 1) / 2)
            if mid * mid > x:
                j = mid - 1
            else:
                i = mid
        return j


if __name__ == '__main__':
    s = Solution()
    # array = [int(i) for i in input().split()]
    param = int(input().strip())
    print(s.my_sqrt(param))
